### DISCUSSION TITLE
Is it possible to add custom operators with custom evaluation functions?

### DISCUSSION THREAD

**Discussion by Lemour-sudo** [react-querybuilder/react-querybuilder#829] [General]
Is it possible to add custom operators with custom evaluation functions:

```
<QueryBuilder
    operators={[
        {
            name: 'strlen',
            label: 'strlen',
            value: 'strlen',
            evaluate: (value: string) => value.length,
        },
        {
            name: 'isOdd',
            label: 'isEven',
            value: 'strlen',
            evaluate: (value: number) => value % 2 === 0,
        },
    ]}
    fields={fields}
    query={query}
    onQueryChange={setQuery}
/>
```

**Comment by jakeboone02** [comment]
You can add any additional properties you want to the operators prop (see [Managing operators](https://react-querybuilder.js.org/docs/tips/managing-operators)), but the `QueryBuilder` component doesn't evaluate data on its own so there's no built-in support for evaluation methods.

Custom operators can be used with any of the [`formatQuery` export formats](https://react-querybuilder.js.org/docs/utils/export), the output of which can be used with a corresponding query library like [JsonLogic](https://react-querybuilder.js.org/docs/utils/export#jsonlogic) or a database platform like [PostgreSQL](https://react-querybuilder.js.org/docs/utils/export#presets).

**Reply by Lemour-sudo** [reply]
Awesome!
Are there commonly used ways or libraries for the evaluations? I am just trying to figure out how I would link the evaluation functions to the queries in the UI component. 

So the default operators too do not have evaluation functions in this library?

**Reply by jakeboone02** [reply]
They do, but only in the context of an evaluation library. Check out any of the "dbquery.*.test.ts" files in this folder to see them in action: https://github.com/react-querybuilder/react-querybuilder/tree/main/packages/react-querybuilder/src/utils/formatQuery.

For local, client-only evaluation (e.g., filtering a data table in the browser), I would recommend JsonLogic. For server-side stuff, PostgreSQL is great but SQLite is fine too. Anything else would need justification for a specific use case.
